common/springUtilities/wupgetLoader.lua

```diff
@@ -0,0 +1,270 @@
+		---@type WupgetInfo
+		local info = infoAccessor(wupget)
+		infoToWupget[info] = wupget
+		wupgetToInfo[wupget] = info
+		if info.parent == nil then
+			resultsList[#resultsList + 1] = wupget
+		else
+			childrenToAdd[info.filename] = wupget
+		end
+	end
+
+	local function infoComp(w1, w2)
+		local info1 = wupgetToInfo[w1]
+		local info2 = wupgetToInfo[w2]
+
+		local l1 = info1.layer
+		local l2 = info2.layer
+		if l1 ~= l2 then
+			return (l1 < l2)
+		end
+		local n1 = info1.name
+		local n2 = info2.name
+		local o1 = orderList[n1]
+		local o2 = orderList[n2]
+		if o1 ~= o2 then
+			return (o1 < o2)
+		else
+			return (n1 < n2)
+		end
+	end
+
+	local function insertChildren(parentIndex, parentInfo)
+		if parentInfo.children and #parentInfo.children > 0 then
+			local currentChildren = {}
+
+			for _, childInfo in ipairs(parentInfo.children) do
+				local child = childrenToAdd[childInfo.filename]
+				if child ~= nil then
+					currentChildren[#currentChildren + 1] =  child
+					childrenToAdd[childInfo.filename] = nil
+				end
+			end
+
+			table.sort(currentChildren, infoComp)
+
+			for childIdx = #currentChildren, 1, -1 do
+				local child = currentChildren[childIdx]
+				table.insert(resultsList, parentIndex + 1, child)
+				insertChildren(parentIndex + 1, wupgetToInfo[child])
+			end
+		end
+	end
+
+	-- sort the wupgets!
+	table.sort(resultsList, infoComp)
+
+	-- iterating backwards is the only way this loop is certain to end
+	-- also makes it much easier to insert into the list
+	-- since new entries will keep pushing existing ones back
+	for idx = #resultsList, 1, -1 do
+		insertChildren(idx, wupgetToInfo[resultsList[idx]])
+	end
+
+	if #childrenToAdd > 0 then
+		Spring.Log('Wupget Loader :: sortWupgets()', LOG.ERROR, "Not all child wupgets were placed into sorted list! Count: " .. #childrenToAdd)
+	end
+
+	return resultsList
+end
+
+-- forward declarations as these functions all rely on each other
+
+---@param fileOrDirPath string
+---@param vfsMode
+---@param loaderCallback fun(filePath:string, parentInfo: WupgetInfo | nil): WupgetInfo
+---@param loadedFilePaths WupgetInfo[] | nil
+---@param parentInfo WupgetInfo | nil
+local function loadFromPath(fileOrDirPath, vfsMode, loaderCallback, loadedFilePaths, parentInfo)
+end
+
+---@param modulePaths string[]
+---@param vfsMode
+---@param loaderCallback fun(filePath:string, parentInfo: WupgetInfo | nil): WupgetInfo
+---@param loadedFilePaths WupgetInfo[] | nil
+---@param parentInfo WupgetInfo | nil
+local function loadFromList(modulePaths, vfsMode, loaderCallback, loadedFilePaths, parentInfo)
+end
+
+---@param dirPath string
+---@param vfsMode
+---@param loaderCallback fun(filePath:string, parentInfo: WupgetInfo | nil): WupgetInfo
+---@param loadedFilePaths WupgetInfo[] | nil
+---@param parentInfo WupgetInfo | nil
+local function loadAllInDir(dirPath, vfsMode, loaderCallback, loadedFilePaths, parentInfo)
+end
+
+
+local recursionDepth = 0
+local MAX_RECURSION_DEPTH = 10
```

Comment:
With `loadedFilePaths` already checking for cycles, is checking recursion depth still necessary?

---

`recursionDepth` is also never reset, so if there's an error during loading and then loading is retried, you might end up with it starting at a non-zero `recursionDepth` value.

Reply:
I think the idea a couple months ago was... to yell at people nesting things that deeply

but since cycles are being protected against someone would need to intentionally make this run forever
I'll take it out